src/Middleware/ConcurrencyLimiter/src/QueuePolicies/QueuePolicy.cs

```diff
@@ -10,49 +10,49 @@ namespace Microsoft.AspNetCore.ConcurrencyLimiter
 {
     internal class QueuePolicy : IQueuePolicy, IDisposable
     {
-        private readonly int _maxConcurrentRequests;
-        private readonly int _requestQueueLimit;
+        private readonly int _maxTotalRequest;
         private readonly SemaphoreSlim _serverSemaphore;
 
         private object _totalRequestsLock = new object();
+
+        private readonly static ValueTask<bool> _falseValueTask = new ValueTask<bool>(false);
```

Comment:
Is there an advantage in caching the ValueTask over creating it where needed?

For Task there is one, as it safes further allocations. 
Here the VT will be copied vs created, I'm not sure what is better and if this is needed.

Reply:
I've done a simple benchmark and returning a stored value task is faster than returning a new instance everytime.

Reply:
Did your benchmark consider that the stored value may not be in the cache (as I would assume for real world scenarios)?

Reply:
I don't know what do you mean by cache but this is my benchmark. Maybe it is *too* simple. What do you think? 

```
public class ValueTaskStoredVsNew
{
    private readonly ValueTask<bool> _storedValueTask = new ValueTask<bool>(false);

    [Benchmark]
    public ValueTask<bool> Stored() => _storedValueTask;

    [Benchmark]
    public ValueTask<bool> New() => new ValueTask<bool>(false);
}

```
| Method |     Mean |     Error |    StdDev | Rank |
|------- |---------:|----------:|----------:|-----:|
| Stored | 11.34 ns | 0.2744 ns | 0.3756 ns |    1 |
|    New | 23.08 ns | 0.5041 ns | 0.6376 ns |    2 |

Reply:
I think this optimization isn't necessary. This would never show up in a real benchmark for request limiting, and as @gfoidl mentioned, the _storedValueTask may not be available in the cache, making this operation slower. The benchmark you are running constantly uses _storedValueTask, meaning it is always in the cache making it quick to retrieve.

Reply:
>  _storedValueTask may not be available in the cache

I don't know which level of cache you are talking about here. Could you please explain a little or point to a documentation? Is it something related to clr?

Also since I'm removing the stored `ValueTask<bool>` here should I do the same for 
https://github.com/aspnet/AspNetCore/blob/master/src/Middleware/ConcurrencyLimiter/src/QueuePolicies/StackPolicy.cs#L24

Reply:
https://en.wikipedia.org/wiki/CPU_cache
`_storedValueTask` is in the RAM, so when accessed the hardware utilizes an optimization to put in the "cache" (via reading an entire cache-line of 64 bytes on x86-systems), so that next accesses to this data are [fast if it is still cached](https://people.eecs.berkeley.edu/~rcs/research/interactive_latency.html). Otherwise the cpu has to read the data again from main memory (RAM) which is a magnitude slower.

In the micro-benchmark the `_storedValueTask` will almost always be in the cache, so reading is fast. In real-world usages `_storedValueTask` will almost always not be in the cache, as other data will "override" the cache, so `_storedValueTask` has to be read from RAM, hence is slower than shown in the micro-benchmark.

The direct creation of the `ValueTask<bool>` doesn't use any memory-hierachy (except stack), therefore I assume this will be faster than the variant with the cached value-task.

Micro-benchmarks are nice, and with benchmarkdotnet easy to do, but the results have to be read and interpreted accordingly. Escpecially in regards to real-usages.
To set up proper benchmarks with cache-eviction is not so trivial, though

>  it's ok to use stored ValueTask in StackPolicy since it's in the default route.

`_bufferLock` and `_freeServerSpots` are accessed, so will be in the cache, therefore it is likely that `_trueTask` will also be in the cache.
Personally I wouldn't use a cache value-task here either. I believe the "pattern" from `Task<bool>` (or `Task<T>` in general) was just copied over. For `Task<T>` caching makes sense, as it avoids an allocation. For `ValueTask<T>` there is no allocation (except it is used as discriminated `Task<T>`, that's a different story), hence the main motivation for caching is gone.

Perf-wise it may be the other way round: caching the value-task may harm perf, as the object gets bigger. Bigger objects is more memory and for field-acces they may be split accross cache-lines, so accesses may become slower.

I write "may", as modern cpus may behave quite different than one might predict according to any theory. Only real-world measurements can show which route is better than another.

Anyway here we talk about micro-optimization that may have no effect in the myriad of other code around them. Readable code that is easy to understand and maintain should have priority here.

Reply:
Thanks a lot for your detailed explanation.

> _bufferLock and _freeServerSpots are accessed, so will be in the cache, therefore it is likely that _trueTask will also be in the cache.

This sentence really did opened my mind :)